src/uploadAssets.js

```diff
@@ -50,5 +63,79 @@ export default async function uploadAssets(
     },
     { apiKey, apiSecret, retryCount: 2 },
   );
+
   return assetsRes.path;
 }
+
+/**
+ * Uploads assets via signed URL
+ *
+ * @param {Buffer|Stream} streamOrBuffer
+ * @param {Object} options
+ * @param {string} options.hash
+ * @param {string} options.endpoint
+ * @param {string} options.apiKey
+ * @param {string} options.apiSecret
+ * @param {Logger} options.logger
+ * @param {string} options.project
+ * @returns {Promise<string>} The URL of the uploaded assets
+ */
+async function uploadAssetsWithSignedUrl(
+  streamOrBuffer,
+  { hash, endpoint, apiKey, apiSecret, logger, project },
+) {
+  // First we need to get the signed URL from Happo.
+  const signedUrlRes = await makeRequest(
+    {
+      url: `${endpoint}/api/snap-requests/assets/${hash}/signed-url`,
+      method: 'GET',
+      json: true,
+    },
+    { apiKey, apiSecret, retryCount: 3 },
+  );
+
+  if (!signedUrlRes) {
+    throw new Error('Failed to get signed URL');
+  }
+
+  // If the asset has already been uploaded, we can return the path now.
+  if (signedUrlRes.path) {
+    logger.info(`${logTag(project)}Reusing existing assets at ${signedUrlRes.path}`);
+    return signedUrlRes.path;
+  }
+
+  // Upload the assets to the signed URL using node's built-in fetch.
+  const putRes = await fetch(signedUrlRes.signedUrl, {
+    method: 'PUT',
+    body: streamOrBuffer,
+    headers: {
+      'Content-Type': 'application/zip',
+    },
+  });
+
+  if (!putRes.ok) {
+    throw new Error(
+      `Failed to upload assets to S3 signed URL: ${putRes.status} ${putRes.statusText}`,
+    );
+  }
+
+  // Finally, we need to tell Happo that we've uploaded the assets.
+  const finalizeRes = await makeRequest(
+    {
+      url: `${endpoint}/api/snap-requests/assets/${hash}/finalize`,
+      method: 'POST',
+      json: true,
+    },
+    { apiKey, apiSecret, retryCount: 3 },
+  );
+
+  return finalizeRes.path;
+}
+
+export default async function uploadAssets(streamOrBuffer, options) {
+  if (process.env.HAPPO_SIGNED_URL) {
+    return uploadAssetsWithSignedUrl(streamOrBuffer, options);
+  }
```

Comment:
I am doing this as an opt-in for now. Once we are confident with this approach, we can roll it out as the default option or perhaps the only option.